### ISSUE TITLE
The return type should not be detected as None

### ISSUE THREAD

**Issue by tonka3000** [astral-sh/ty#705]
### Summary

In the below scenario ty report that the return type is `Version|None` but the code can only be `Version`.
```py
class Version:
    def __init__(self):
        self.foo = 1


_ver: Version | None = None


def get_version() -> Version:
    global _ver
    if _ver is not None:
        return _ver # <=== ty error here
    _ver = Version()
    return _ver
```

ty report **error[invalid-return-type] <filename>: Return type does not match returned value: expected `Version`, found `Version | None`**

### Version

0.0.1-alpha.12 (f7446a6ee 2025-06-25)

**Comment by AlexWaygood** [comment]
Thanks! This is #311
